packages/jupyter-ai/jupyter_ai/chat_handlers/default.py

```diff
@@ -130,19 +134,32 @@ async def process_message(self, message: HumanChatMessage):
             # implement streaming, as `astream()` defaults to yielding `_call()`
             # when `_stream()` is not implemented on the LLM class.
             metadata_handler = MetadataCallbackHandler()
-            async for chunk in self.llm_chain.astream(
+            chunk_generator = self.llm_chain.astream(
                 inputs,
                 config={
                     "configurable": {"last_human_msg": message},
                     "callbacks": [metadata_handler],
                 },
-            ):
+            )
+            async for chunk in chunk_generator:
                 if not received_first_chunk:
                     # when receiving the first chunk, close the pending message and
                     # start the stream.
                     self.close_pending(pending_message)
                     stream_id = self._start_stream(human_msg=message)
                     received_first_chunk = True
+                    self.message_interrupted[stream_id] = asyncio.Event()
+
+                if self.message_interrupted[stream_id].is_set():
+                    try:
+                        # notify the model provider that streaming was interrupted
+                        # (this is essential to allow the model to stop generating)
+                        chunk_generator.athrow(GenerationInterrupted())
+                    except GenerationInterrupted:
```

Comment:
Since this is the non-trivial part, here are some snippets to demonstrate how it works - I hope it helps with the review :)

![image](https://github.com/user-attachments/assets/4334f023-2ac5-4a8e-bd57-de50e54b38d1)

![image](https://github.com/user-attachments/assets/993a0c58-d217-4c66-b200-1a5fc1688924)